File: frontend/src/constants/pricing.ts
/**
 * Pricing constants for different AI models.
 * Prices are in USD per 1M tokens (input, output).
 */

import { ModelType } from '../types';

export interface ModelPricing {
  inputPricePerMillion: number;
  outputPricePerMillion: number;
}

export const MODEL_PRICING: Record<ModelType, ModelPricing> = {
  // Claude Models
  'opus-4.5': {
    inputPricePerMillion: 15.00,
    outputPricePerMillion: 75.00,
  },
  'sonnet-4.5': {
    inputPricePerMillion: 3.00,
    outputPricePerMillion: 15.00,
  },
  'haiku-4.5': {
    inputPricePerMillion: 0.25,
    outputPricePerMillion: 1.25,
  },
  // Gemini Models
  'gemini-3-pro': {
    inputPricePerMillion: 1.25,
    outputPricePerMillion: 5.00,
  },
  'gemini-3-flash': {
    inputPricePerMillion: 0.075,
    outputPricePerMillion: 0.30,
  },
};

/**
 * Calculate cost based on model and token usage.
 */
export function calculateModelCost(
  model: ModelType,
  tokens: { input: number; output: number }
): number {
  const pricing = MODEL_PRICING[model];
  if (!pricing) return 0;

  const inputCost = (tokens.input / 1_000_000) * pricing.inputPricePerMillion;
  const outputCost = (tokens.output / 1_000_000) * pricing.outputPricePerMillion;

  return inputCost + outputCost;
}

/**
 * Estimate cost for a full workflow based on selected models.
 * Uses average token estimates.
 */
export function estimateWorkflowCost(
  modelPlan: ModelType,
  modelImplement: ModelType,
  modelTest: ModelType,
  modelReview: ModelType
): number {
  // Estimativas médias baseadas em histórico
  const avgTokens = {
    plan: { input: 5000, output: 2000 },
    implement: { input: 10000, output: 5000 },
    test: { input: 8000, output: 3000 },
    review: { input: 7000, output: 2500 },
  };

  return (
    calculateModelCost(modelPlan, avgTokens.plan) +
    calculateModelCost(modelImplement, avgTokens.implement) +
    calculateModelCost(modelTest, avgTokens.test) +
    calculateModelCost(modelReview, avgTokens.review)
  );
}

File: backend/src/config/pricing.py
"""Pricing configuration for different AI models."""

from decimal import Decimal
from typing import Dict, Tuple

# Preços em USD por 1M tokens (input, output)
MODEL_PRICING: Dict[str, Tuple[Decimal, Decimal]] = {
    # Claude 4.5 models
    "opus-4.5": (Decimal("15.00"), Decimal("75.00")),
    "sonnet-4.5": (Decimal("3.00"), Decimal("15.00")),
    "haiku-4.5": (Decimal("0.25"), Decimal("1.25")),

    # Gemini 3 models
    "gemini-3-pro": (Decimal("1.25"), Decimal("5.00")),
    "gemini-3-flash": (Decimal("0.075"), Decimal("0.30")),
}


def calculate_cost(model: str, input_tokens: int, output_tokens: int) -> Decimal:
    """Calcula custo baseado no modelo e tokens.

    Args:
        model: Nome do modelo utilizado
        input_tokens: Quantidade de tokens de entrada
        output_tokens: Quantidade de tokens de saída

    Returns:
        Custo total em USD como Decimal
    """
    if model not in MODEL_PRICING:
        return Decimal("0")

    input_price, output_price = MODEL_PRICING[model]

    # Converter tokens para milhões e calcular
    input_cost = (Decimal(input_tokens) / 1_000_000) * input_price
    output_cost = (Decimal(output_tokens) / 1_000_000) * output_price

    return input_cost + output_cost
